[6.18] Track btrfs patches

kakra opened the pull request:
Export patch series: https://github.com/kakra/linux/pull/40.patch

[![wakatime](https://wakatime.com/badge/github/kakra/linux.svg)](https://wakatime.com/badge/github/kakra/linux)

# btrfs: tiered allocation hints and queue-based read balancing

**Special Thanks** to @Forza-tng for extensive testing, feedback, and maintaining the documentation guide:
👉 **[Btrfs Allocator Hints and Read Policies Guide by Forza-tng](https://wiki.tnonline.net/w/Btrfs/Allocator_Hints)**

---

This PR introduces a set of patches to improve Btrfs performance and flexibility in heterogeneous storage environments (mixed SSD/HDD, tiered storage, bcache).

## 1. Allocator Hints (Data Placement)

Allows preferring specific devices for data or metadata allocations. This works by storing a hint in the persistent device item on-disk.

- **Essential for tiered setups:** Force metadata onto SSDs (for speed) while keeping bulk data on HDDs.
- **Graceful removal:** Mark devices to accept no new allocations, allowing to drain them naturally or via balance without racing against new writes.

## 2. Read Balancing Policies

Extends Btrfs RAID1 read balancing with queue-aware read policies. The standard PID-based policy is often insufficient for mixed-device pools, high-IOPS workloads, or devices with unstable latency.

- **`pid`**: (Default) Static hashing by process ID. Good for simple setups.
- **`round-robin`**: Distributes reads equally. Good for aggregate throughput on identical disks.
- **`queue`**: **(Old recommendation)** Routes requests to the device with the fewest in-flight requests.
  - Adapts instantly to current device load.
  - Breaks equal-depth ties by mirror age, so idle mirrors are rotated more fairly instead of implicitly preferring the first candidate.
  - Works well for healthy devices and general mixed-load systems.
- **`queue-adaptive`**: **(Recommended for latency-sensitive desktops or mixed/aging devices, still experimental)** Keeps the `queue` policy as the primary selector, but adds a peer-relative latency guard.
  - Temporarily avoids mirror candidates whose recent read latency is much worse than their current peer.
  - Helps with transient SSD/NVMe latency spikes, slow HDD areas, weak sectors, and mixed pools where one mirror is temporarily much slower.
  - Still allows a slower mirror to absorb overflow when the faster peer is sufficiently busier.
  - Can reduce desktop and gaming stalls caused by latency-sick mirror members.
- **`devid`**: Pin reads to a specific device ID (mostly for testing).

On fully healthy and evenly performing devices, `queue-adaptive` should stay close to `queue`. Its benefit is expected to be most visible under mixed load, mixed devices, or devices with temporary or localized latency spikes.

*(Note: Older global latency-based policies were dropped in favor of the queue-based design. `queue-adaptive` uses latency only as a peer-relative guard on top of `queue`, not as the primary selector.)*

## 3. Decoupling from Experimental Status

**Important:** Upstream Kernels (6.13+) use `CONFIG_BTRFS_EXPERIMENTAL` to gate various unstable work-in-progress features. To allow using the allocator hints and read policies without enabling potentially unstable upstream code, these features have been moved out of the experimental gate.

**Recommendation:** Remove the line `CONFIG_BTRFS_EXPERIMENTAL` from your `.config` before running `make oldconfig`. The build system will then prompt you specifically for the new options (`CONFIG_BTRFS_ALLOCATOR_HINTS`, `CONFIG_BTRFS_READ_POLICIES`), allowing you to enable them safely without turning on other experimental Btrfs features.

---

## Quickstart Guide

### Setting Allocator Hints

1. Enable `CONFIG_BTRFS_ALLOCATOR_HINTS` in kernel config.
2. Run `btrfs device usage /mnt/path` to identify your device IDs.
3. Set the hint via sysfs (this persists on-disk, no udev rules needed!):
   `echo <TYPE> | sudo tee /sys/fs/btrfs/<UUID>/devinfo/<DEVID>/type`

**Available Types:**

- `0`: **Prefer data** (Default for HDDs).
- `1`: **Prefer metadata** (Recommended for SSDs/NVMe).
- `2`: Metadata only (Use with caution).
- `3`: Data only (Use with caution).
- `4`: **None preferred** (Avoids new allocations, useful to drain a drive).
- **Added:** `5`: **None** (Strictly prevents ANY new allocation, useful for parallel device remove).

*After changing hints, a rebalance of metadata/data is required to move existing extents to their preferred location.*

### Enabling Read Policy

1. Enable `CONFIG_BTRFS_READ_POLICIES` in kernel config.
2. Set boot parameter: `btrfs.read_policy=queue-adaptive` or `btrfs.read_policy=queue`
3. Or switch at runtime:
   - `echo queue-adaptive | sudo tee /sys/fs/btrfs/<UUID>/read_policy`
   - `echo queue | sudo tee /sys/fs/btrfs/<UUID>/read_policy`

---

## Diagnostic Statistics

Adds per-device read statistics to `/sys/fs/btrfs/<UUID>/devinfo/<DEVID>/read_stats`.

```
ios %lu wait %llu avg %llu age %llu ignored %llu health avg %llu ago %llu score %llu vetoed %llu overflow %llu
```

- **`ios`**: Total read I/O count.
- **`wait`**: Total accumulated read wait time in nanoseconds.
- **`avg`**: Cumulative average read latency in nanoseconds.
- **`age`**: Fairness counter. Increments while the device is a read candidate but not selected. Resets to 0 when selected.
- **`ignored`**: Total count of times this device was a candidate but skipped.
- **`health avg`**: Recent windowed average read latency in nanoseconds, used by `queue-adaptive`.
- **`ago`**: Seconds since `health avg` was last successfully refreshed.
- **`score`**: Ratio between cumulative average latency and recent health latency. Values above 100 mean the recent window is faster than the cumulative average; values below 100 mean it is slower.
- **`vetoed`**: Number of times `queue-adaptive` rejected this device because its recent latency looked much worse than its peer.
- **`overflow`**: Number of times `queue-adaptive` still selected this slower-looking device because the healthier peer was already sufficiently busier.

---

## Benchmark Results

The following benchmarks (based on kernel 6.12 LTS) compare the new policies against the defaults. Tests were performed on a mixed HDD RAID10 array with bcache, comparing an idle system vs. a system under heavy background load (defrag).

**`queue`** proved to be the superior all-rounder, effectively isolating foreground workloads from background noise. **`queue-adaptive`** hasn't been benchmarked yet. The following are the results from the old version:

### Scenario: No Background Load

| Policy | RandRead 4k QD1 (Lat) | RandRead 4k QD32 (IOPS) | SeqRead 1M (BW) |
| :--- | :--- | :--- | :--- |
| `pid` | 65 IOPS | 537 IOPS | 261 MiB/s |
| `round-robin` | 241 IOPS | 1180 IOPS | 231 MiB/s |
| `latency-rr`* | 702 IOPS | 2477 IOPS | 240 MiB/s |
| **`queue`** | **1181 IOPS** | **3647 IOPS** | **272 MiB/s** |

### Scenario: Heavy Background Load (Defrag)

| Policy | RandRead 4k QD1 (Lat) | RandRead 4k QD32 (IOPS) | SeqRead 1M (BW) |
| :--- | :--- | :--- | :--- |
| `pid` | ~0 IOPS (Stalled) | 505 IOPS | 121 MiB/s |
| `round-robin` | 38 IOPS | 717 IOPS | 126 MiB/s |
| `latency-rr`* | 585 IOPS | 1562 IOPS | 235 MiB/s |
| **`queue`** | **967 IOPS** | **2437 IOPS** | **247 MiB/s** |

*(`latency-rr` was an experimental hybrid policy used during testing, superseded by `queue` due to better performance and simplicity, `queue-adaptive` not yet included in these results)*

---

## Changes in this version (Kernel 6.18 Port)

- **Ported to Linux 6.18.**
- **Refactored Kconfig:** Features are individually selectable and moved out of "Experimental".
- **Robustness:** Fixed potential NULL pointer dereferences in stats tracking during mount/unmount and race conditions in round-robin calculation.
- **Simplified:** Dropped older global/EMA-based latency policies. The new queue-adaptive policy uses latency only as a peer-relative guard on top of queue.

---

## Status

- Intended for **advanced users** and **maintained out-of-tree**.
- **Not expected** to be accepted upstream in current form.
- **Allocator hints** are experimental but used in production by maintainers/testers.
- Use `btrfs filesystem usage -T` after enabling hints to **confirm metadata placement**.

---

## Upstream outlook

This patchset **is not currently an upstream submission**. It serves as a **practical implementation and testbed for tiered Btrfs allocation and read policy ideas**. Some concepts may later be replaced by upstream-native mechanisms.

---

## FAQ: Why is this not upstream?

### 1. Allocator Hints

The allocator hint patches (originally developed by **Goffredo Baroncelli**, now maintained here) have been discussed on the mailing list but were not merged for design reasons:

- **Free Space Calculation (`df`):** Btrfs calculates available space assuming any chunk can be allocated on any device (respecting RAID profiles). Restricting allocations via hints makes this calculation unreliable. Tools might report free space while Btrfs returns `ENOSPC` (No space left on device) because the allowed devices for a specific chunk type are full, even if other devices are empty.
- **Maintenance:** The original author ceased updates for newer kernels; this repository bridges that gap.

**Compatibility Note:** This patch reuses the existing (unused) `type` field in the device item on disk. It does **not** change the on-disk format version. Unpatched kernels simply ignore the value, ensuring data remains accessible (though allocation preferences will be lost until booted with a patched kernel).

### 2. Read Policies (`queue` and `queue-adaptive`)

The new `queue` and `queue-adaptive` policies are an experimental addition in this patch set. It is unlikely to be accepted upstream in its current form due to **Layer Violation**:

- The filesystem layer (Btrfs) directly accesses internal block-layer statistics (in-flight queue depth) to make routing decisions. The Linux kernel generally enforces strict separation between these subsystems.
- However, benchmarks show this cross-layer optimization yields significant performance gains in mixed setups, justifying its inclusion here.

kakra committed:
btrfs: add allocator_hint for no allocation preferred

This is useful where you want to prevent new allocations of chunks on a
disk which is going to be removed from the pool anyways, e.g. due to
bad blocks or because it's slow.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add allocator_hint to disable allocation completely

This is useful where you want to prevent new allocations of chunks to
a set of multiple disks which are going to be removed from the pool.
This acts as a multiple `btrfs dev remove` on steroids that can remove
multiple disks in parallel without moving data to disks which would be
removed in the next round. In such cases, it will avoid moving the
same data multiple times, and thus avoid placing it on potentially bad
disks.

Thanks to @Zygo for the explanation and suggestion.

Link: https://github.com/kdave/btrfs-progs/issues/907#issuecomment-2520897104
Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add in-flight queue read policy

Select the preferred stripe based on the mirror with the least
in-flight requests.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add in-flight queue read policy

Select the preferred stripe based on the mirror with the least
in-flight requests.

v2:
- Skip missing mirrors before reading their in-flight statistics so the
  policy remains safe on degraded arrays.
- Initialize the fallback stripe from the current mirror set instead of
  assuming stripe zero.

Signed-off-by: Kai Krakow <kai@kaishome.de>

fdmanana committed:
btrfs: avoid RCU stall during backref list construction

The following kernel message may be logged if `add_inline_refs()` or
`add_keyed_refs()` block for too long:

> kernel: rcu: INFO: rcu_sched self-detected stall on CPU
> kernel: rcu:         10-....: (2100 ticks this GP) idle=0494/1/0x4000000000000000 softirq=164826140/164826187 fqs=1052
> kernel: rcu:         (t=2100 jiffies g=358306033 q=2241752 ncpus=16)
> kernel: CPU: 10 UID: 0 PID: 1524681 Comm: map_0x178e45670 Not tainted 6.12.21-gentoo #1
> kernel: Hardware name: Red Hat KVM, BIOS 0.5.1 01/01/2011
> kernel: RIP: 0010:btrfs_get_64+0x65/0x110
> kernel: Code: d3 ed 48 8b 4f 70 48 8b 31 83 e6 40 74 11 0f b6 49 40 41 bc 00 10 00 00 49 d3 e4 49 83 ec 01 4a 8b 5c ed 70 49 21 d4 45 89 c9 <48> 2b 1d 7c 99 09 01 49 01 c1 8b 55 08 49 8d 49 08 44 8b 75 0c 48
> kernel: RSP: 0018:ffffbb7ad531bba0 EFLAGS: 00000202
> kernel: RAX: 0000000000001f15 RBX: fffff437ea382200 RCX: fffff437cb891200
> kernel: RDX: 000001922b68df2a RSI: 0000000000000000 RDI: ffffa434c3e66d20
> kernel: RBP: ffffa434c3e66d20 R08: 000001922b68c000 R09: 0000000000000015
> kernel: R10: 6c0000000000000a R11: 0000000009fe7000 R12: 0000000000000f2a
> kernel: R13: 0000000000000001 R14: ffffa43192e6d230 R15: ffffa43160c4c800
> kernel: FS:  000055d07085e6c0(0000) GS:ffffa4452bc80000(0000) knlGS:0000000000000000
> kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
> kernel: CR2: 00007fff204ecfc0 CR3: 0000000121a0b000 CR4: 00000000001506f0
> kernel: DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
> kernel: DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
> kernel: Call Trace:
> kernel:  <IRQ>
> kernel:  ? rcu_dump_cpu_stacks+0xd3/0x100
> kernel:  ? rcu_sched_clock_irq+0x4ff/0x920
> kernel:  ? update_process_times+0x6c/0xa0
> kernel:  ? tick_nohz_handler+0x82/0x110
> kernel:  ? tick_do_update_jiffies64+0xd0/0xd0
> kernel:  ? __hrtimer_run_queues+0x10b/0x190
> kernel:  ? hrtimer_interrupt+0xf1/0x200
> kernel:  ? __sysvec_apic_timer_interrupt+0x44/0x50
> kernel:  ? sysvec_apic_timer_interrupt+0x60/0x80
> kernel:  </IRQ>
> kernel:  <TASK>
> kernel:  ? asm_sysvec_apic_timer_interrupt+0x16/0x20
> kernel:  ? btrfs_get_64+0x65/0x110
> kernel:  find_parent_nodes+0x1b84/0x1dc0
> kernel:  btrfs_find_all_leafs+0x31/0xd0
> kernel:  ? queued_write_lock_slowpath+0x30/0x70
> kernel:  iterate_extent_inodes+0x6f/0x370
> kernel:  ? update_share_count+0x60/0x60
> kernel:  ? extent_from_logical+0x139/0x190
> kernel:  ? release_extent_buffer+0x96/0xb0
> kernel:  iterate_inodes_from_logical+0xaa/0xd0
> kernel:  btrfs_ioctl_logical_to_ino+0xaa/0x150
> kernel:  __x64_sys_ioctl+0x84/0xc0
> kernel:  do_syscall_64+0x47/0x100
> kernel:  entry_SYSCALL_64_after_hwframe+0x4b/0x53
> kernel: RIP: 0033:0x55d07617eaaf
> kernel: Code: 00 48 89 44 24 18 31 c0 48 8d 44 24 60 c7 04 24 10 00 00 00 48 89 44 24 08 48 8d 44 24 20 48 89 44 24 10 b8 10 00 00 00 0f 05 <89> c2 3d 00 f0 ff ff 77 18 48 8b 44 24 18 64 48 2b 04 25 28 00 00
> kernel: RSP: 002b:000055d07085bc20 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
> kernel: RAX: ffffffffffffffda RBX: 000055d0402f8550 RCX: 000055d07617eaaf
> kernel: RDX: 000055d07085bca0 RSI: 00000000c038943b RDI: 0000000000000003
> kernel: RBP: 000055d07085bea0 R08: 00007fee46c84080 R09: 0000000000000000
> kernel: R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000000003
> kernel: R13: 000055d07085bf80 R14: 000055d07085bf48 R15: 000055d07085c0b0
> kernel:  </TASK>

The RCU stall could be because there's a large number of backrefs for
some extents and we're spending too much time looping over them
without ever yielding the cpu.

Avoid the stall warning by adding `conf_resched()`.

Link: https://lore.kernel.org/linux-btrfs/CAMthOuP_AE9OwiTQCrh7CK73xdTZvHsLTB1JU2WBK6cCc05JYg@mail.gmail.com/T/#md2e3504a1885c63531f8eefc70c94cff571b7a72
Signed-off-by: Kai Krakow <kk@netactive.de>

kakra committed:
btrfs: avoid RCU stall during backref list construction

The following kernel message may be logged if `add_inline_refs()` or
`add_keyed_refs()` block for too long:

> kernel: rcu: INFO: rcu_sched self-detected stall on CPU
> kernel: rcu:         10-....: (2100 ticks this GP) idle=0494/1/0x4000000000000000 softirq=164826140/164826187 fqs=1052
> kernel: rcu:         (t=2100 jiffies g=358306033 q=2241752 ncpus=16)
> kernel: CPU: 10 UID: 0 PID: 1524681 Comm: map_0x178e45670 Not tainted 6.12.21-gentoo #1
> kernel: Hardware name: Red Hat KVM, BIOS 0.5.1 01/01/2011
> kernel: RIP: 0010:btrfs_get_64+0x65/0x110
> kernel: Code: d3 ed 48 8b 4f 70 48 8b 31 83 e6 40 74 11 0f b6 49 40 41 bc 00 10 00 00 49 d3 e4 49 83 ec 01 4a 8b 5c ed 70 49 21 d4 45 89 c9 <48> 2b 1d 7c 99 09 01 49 01 c1 8b 55 08 49 8d 49 08 44 8b 75 0c 48
> kernel: RSP: 0018:ffffbb7ad531bba0 EFLAGS: 00000202
> kernel: RAX: 0000000000001f15 RBX: fffff437ea382200 RCX: fffff437cb891200
> kernel: RDX: 000001922b68df2a RSI: 0000000000000000 RDI: ffffa434c3e66d20
> kernel: RBP: ffffa434c3e66d20 R08: 000001922b68c000 R09: 0000000000000015
> kernel: R10: 6c0000000000000a R11: 0000000009fe7000 R12: 0000000000000f2a
> kernel: R13: 0000000000000001 R14: ffffa43192e6d230 R15: ffffa43160c4c800
> kernel: FS:  000055d07085e6c0(0000) GS:ffffa4452bc80000(0000) knlGS:0000000000000000
> kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
> kernel: CR2: 00007fff204ecfc0 CR3: 0000000121a0b000 CR4: 00000000001506f0
> kernel: DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
> kernel: DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
> kernel: Call Trace:
> kernel:  <IRQ>
> kernel:  ? rcu_dump_cpu_stacks+0xd3/0x100
> kernel:  ? rcu_sched_clock_irq+0x4ff/0x920
> kernel:  ? update_process_times+0x6c/0xa0
> kernel:  ? tick_nohz_handler+0x82/0x110
> kernel:  ? tick_do_update_jiffies64+0xd0/0xd0
> kernel:  ? __hrtimer_run_queues+0x10b/0x190
> kernel:  ? hrtimer_interrupt+0xf1/0x200
> kernel:  ? __sysvec_apic_timer_interrupt+0x44/0x50
> kernel:  ? sysvec_apic_timer_interrupt+0x60/0x80
> kernel:  </IRQ>
> kernel:  <TASK>
> kernel:  ? asm_sysvec_apic_timer_interrupt+0x16/0x20
> kernel:  ? btrfs_get_64+0x65/0x110
> kernel:  find_parent_nodes+0x1b84/0x1dc0
> kernel:  btrfs_find_all_leafs+0x31/0xd0
> kernel:  ? queued_write_lock_slowpath+0x30/0x70
> kernel:  iterate_extent_inodes+0x6f/0x370
> kernel:  ? update_share_count+0x60/0x60
> kernel:  ? extent_from_logical+0x139/0x190
> kernel:  ? release_extent_buffer+0x96/0xb0
> kernel:  iterate_inodes_from_logical+0xaa/0xd0
> kernel:  btrfs_ioctl_logical_to_ino+0xaa/0x150
> kernel:  __x64_sys_ioctl+0x84/0xc0
> kernel:  do_syscall_64+0x47/0x100
> kernel:  entry_SYSCALL_64_after_hwframe+0x4b/0x53
> kernel: RIP: 0033:0x55d07617eaaf
> kernel: Code: 00 48 89 44 24 18 31 c0 48 8d 44 24 60 c7 04 24 10 00 00 00 48 89 44 24 08 48 8d 44 24 20 48 89 44 24 10 b8 10 00 00 00 0f 05 <89> c2 3d 00 f0 ff ff 77 18 48 8b 44 24 18 64 48 2b 04 25 28 00 00
> kernel: RSP: 002b:000055d07085bc20 EFLAGS: 00000246 ORIG_RAX: 0000000000000010
> kernel: RAX: ffffffffffffffda RBX: 000055d0402f8550 RCX: 000055d07617eaaf
> kernel: RDX: 000055d07085bca0 RSI: 00000000c038943b RDI: 0000000000000003
> kernel: RBP: 000055d07085bea0 R08: 00007fee46c84080 R09: 0000000000000000
> kernel: R10: 0000000000000000 R11: 0000000000000246 R12: 0000000000000003
> kernel: R13: 000055d07085bf80 R14: 000055d07085bf48 R15: 000055d07085c0b0
> kernel:  </TASK>

The RCU stall could be because there's a large number of backrefs for
some extents and we're spending too much time looping over them
without ever yielding the cpu.

Avoid the stall warning by adding `conf_resched()`.

Link: https://lore.kernel.org/linux-btrfs/CAMthOuP_AE9OwiTQCrh7CK73xdTZvHsLTB1JU2WBK6cCc05JYg@mail.gmail.com/T/#md2e3504a1885c63531f8eefc70c94cff571b7a72
Suggested-by: Filipe Manana <fdmanana@kernel.org>
Signed-off-by: Kai Krakow <kk@netactive.de>

kakra committed:
btrfs: add new Kconfig option for btrfs allocator hints

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add flags to give a hint to the chunk allocator

Add the following flags to give a hint about which chunk should be
allocated on which a disk.

The following flags are created:

- BTRFS_DEV_ALLOCATION_PREFERRED_DATA
  preferred data chunk, but metadata chunk allowed
- BTRFS_DEV_ALLOCATION_PREFERRED_METADATA
  preferred metadata chunk, but data chunk allowed
- BTRFS_DEV_ALLOCATION_METADATA_ONLY
  only metadata chunk allowed
- BTRFS_DEV_ALLOCATION_DATA_ONLY
  only data chunk allowed

Co-authored-by: Goffredo Baroncelli <kreijack@inwid.it>
Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: export dev_item.type in /sys/fs/btrfs/<uuid>/devinfo/<devid>/type

Co-authored-by: Goffredo Baroncelli <kreijack@inwind.it>
Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add allocator_hint mode

When this mode is enabled, the chunk allocation policy is modified as
follows:

Each disk may have a different tag:
- BTRFS_DEV_ALLOCATION_PREFERRED_METADATA
- BTRFS_DEV_ALLOCATION_METADATA_ONLY
- BTRFS_DEV_ALLOCATION_DATA_ONLY
- BTRFS_DEV_ALLOCATION_PREFERRED_DATA (default)

Where:
- ALLOCATION_PREFERRED_X means that it is preferred to use this disk
  for the X chunk type (the other type may be allowed when the space is
  low)
- ALLOCATION_X_ONLY means that it is used *only* for the X chunk type.
  This means also that it is a preferred choice.

Each time the allocator allocates a chunk of type X, first it takes the
disks tagged as ALLOCATION_X_ONLY or ALLOCATION_PREFERRED_X.

If the space is not enough, it uses also the disks tagged as
ALLOCATION_METADATA_ONLY.

If the space is not enough, it uses also the other disks, with the
exception of the one marked as ALLOCATION_PREFERRED_Y, where Y is the
other type of chunk (i.e. not X).

Co-authored-by: Goffredo Baroncelli <kreijack@inwind.it>
Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add io read stats per device to devinfo

This adds read stats per device to devinfo to evaluate the effects of
different read policies better.

This adds a new file /sys/fs/btrfs/BTRFS-UUID/devinfo/ID/read_stats.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: move read policies out of experimental

Read policies seem safe and stable enough to move it out of the
experimental feature set. This allows us to add more policies without
forcing users to enable the full experimental feature set.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: move read policies out of experimental

Read policies seem safe and stable enough to move it out of the
experimental feature set. This allows us to add more policies without
forcing users to enable the full experimental feature set.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: change the DEV_ITEM 'type' field via sysfs

v2: Adds a check to prevent modification while the file system is still mounting.

Todo:

- Transactions should not be triggered from sysfw writes, see:
  https://lore.kernel.org/linux-btrfs/20251213200920.1808679-1-kai@kaishome.de/

Link: https://github.com/kakra/linux/pull/36#issuecomment-3406301805
Reported-by: Eli Venter <eli@genedx.com>
Co-authored-by: Goffredo Baroncelli <kreijack@inwind.it>
Signed-off-by: Kai Krakow <kai@kaishome.de>

CHN-beta commented:
I tried this patch with CachyOS kernel 6.18.0. After recompiling and rebooting, ` /sys/fs/btrfs/<UUID>/devinfo/<DEVID>/type` did not exist. I can confirm that the patch is applied, and the corresponding kernel options are set.

```
$ zcat /proc/config.gz | grep -i btrfs
CONFIG_BTRFS_FS=m
CONFIG_BTRFS_FS_POSIX_ACL=y
# CONFIG_BTRFS_FS_RUN_SANITY_TESTS is not set
# CONFIG_BTRFS_DEBUG is not set
# CONFIG_BTRFS_ASSERT is not set
CONFIG_BTRFS_ALLOCATOR_HINTS=y
# CONFIG_BTRFS_PER_DEVICE_IO_STATS is not set
CONFIG_BTRFS_READ_POLICIES=y
CONFIG_BTRFS_EXPERIMENTAL=y
$ ls /sys/fs/btrfs/f2dfd4a4-276d-4451-999e-a39457f032b5/devinfo/2
error_stats  fsid  in_fs_metadata  missing  replace_target  scrub_speed_max  writeable
```

Is there anything I am missing?

Forza-tng commented:
> I tried this patch with CachyOS kernel 6.18.0. After recompiling and rebooting, ` /sys/fs/btrfs/<UUID>/devinfo/<DEVID>/type` did not exist. I can confirm that the patch is applied, and the corresponding kernel options are set.
> 
> ```
> $ zcat /proc/config.gz | grep -i btrfs
> CONFIG_BTRFS_FS=m
> CONFIG_BTRFS_FS_POSIX_ACL=y
> # CONFIG_BTRFS_FS_RUN_SANITY_TESTS is not set
> # CONFIG_BTRFS_DEBUG is not set
> # CONFIG_BTRFS_ASSERT is not set
> CONFIG_BTRFS_ALLOCATOR_HINTS=y
> # CONFIG_BTRFS_PER_DEVICE_IO_STATS is not set
> CONFIG_BTRFS_READ_POLICIES=y
> CONFIG_BTRFS_EXPERIMENTAL=y
> $ ls /sys/fs/btrfs/f2dfd4a4-276d-4451-999e-a39457f032b5/devinfo/2
> error_stats  fsid  in_fs_metadata  missing  replace_target  scrub_speed_max  writeable
> ```
> 
> Is there anything I am missing?

Thanks for the reporty. I can confirm the same.

kakra commented:
Thanks for the report. Will fix...

kakra commented:
@CHN-beta @Forza-tng Thanks for reporting and confirming. This was actually a bug I introduced when I made allocator hints configurable via `make menuconfig`: I used the wrong definition in the C code, the allocator hints actually never compiled and were not active (it thus also slipped through my compile tests which showed two syntax issues now). I never verified if the type fields still existed, I'll keep this in mind for the future. Sorry.

**Important:** This means, whoever used the 6.18 patches until now, never had allocator hints enabled since. Please use the new patches, then verify that `devinfo/*/type` exists (it will still have your original type value). If it exists, it means the patch is working now. But you'll need to check if your btrfs moved meta data to the slow devices:
```
# btrfs filesystem usage -T {BTRFS-MOUNT-PATH}
...
                  Data    Metadata System
Id Path           RAID1   RAID1    RAID1    Unallocated Total     Slack
-- -------------- ------- -------- -------- ----------- --------- -------
 1 /dev/bcache2   2.51TiB        -        -     1.12TiB   3.63TiB 3.50KiB
 2 /dev/bcache0   2.52TiB        -        -     1.12TiB   3.63TiB 3.50KiB
 4 /dev/nvme0n1p2       - 86.00GiB 32.00MiB    41.97GiB 128.00GiB       -
 6 /dev/nvme1n1p2       - 86.00GiB 32.00MiB    41.97GiB 128.00GiB       -
 7 /dev/bcache3   2.52TiB        -        -     1.12TiB   3.64TiB       -
 8 /dev/bcache1   2.50TiB        -        -     1.14TiB   3.64TiB       -
-- -------------- ------- -------- -------- ----------- --------- -------
   Total          5.03TiB 86.00GiB 32.00MiB     4.57TiB  14.79TiB 7.00KiB
   Used           4.92TiB 30.35GiB  1.05MiB
```

If it lists devices with unexpected meta data, take note of the affected device IDs, then run a meta data balance filtered for device ID (separate each ID by spaces):
```bash
for ID in {SLOW_DEV_IDs}; do btrfs balance start -mdevid=$ID --enqueue {BTRFS-MOUNT-PATH}; done
```

E.g., run `for ID in 1 7; do ...` if your meta data ended up unwanted on device ID 1 and 7. Balance will then rewrite all meta chunks on device 1 and 7, effectively re-allocating it from the fast dedicated devices, without touching existing meta data chunks on other devices. It should be a fast and safe operation.

Thanks.

CHN-beta commented:
>  I never verified if the type fields still existed, I'll keep this in mind for the future. Sorry.

Please don't worry about it. Everyone makes mistakes sometimes, and this one didn't actually cause any damage. Thank you for your contribution!

kakra commented:
Updated the branch to improve the style of some if statements.

kakra committed:
GITHUB: btrfs: add build workflow

kakra committed:
GITHUB: btrfs: add build workflow

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra commented:
Added a Github workflow to automatically compile and build-check the patches.

kakra committed:
btrfs: add queue-adaptive read policy

Add a new RAID1 read policy, queue-adaptive, selected independently of
the existing queue policy.

queue picks the mirror with the fewest in-flight requests and uses
last_io_age to break equal-depth ties. This works well on healthy
devices but can still flap onto a latency-sick device: a stuck or slow
device can show a lower in-flight count simply because its last request
has not completed yet, so queue repeatedly picks it whenever the healthy
device has one extra in-flight request.

queue-adaptive keeps queue's in-flight comparison and shared age
tie-break as the primary selector, but adds a coarse latency veto. For
each read, it compares only the actual current mirror candidates' cached
windowed latency averages. A device more than 4x worse than the best
current peer is treated as slow for that call.

The slow in-flight winner is rejected in favor of the best non-slow
candidate only if its in-flight lead is not more than the guard margin.
This keeps the latency guard from turning into a hard device ban: a slow
mirror can still absorb overflow once the healthier peer's queue backs
up enough.

When per-device IO stats are enabled, read_stats also reports the age of
the last successful health refresh, a health score relative to the
lifetime average, how often a device was vetoed by this guard and how
often it was still selected as slow because its in-flight lead exceeded
the guard margin. These counters are diagnostic only and do not feed
back into selection.

v2:
- Refresh health samples only while queue-adaptive is active, avoiding
  block-stat and division overhead for all other read policies. The ago
  field exposes when the retained sample was last refreshed.
- Skip missing mirrors before reading health or in-flight state so the
  policy remains safe on degraded arrays.
- Clarify that score compares the current window with the device's own
  lifetime history and is not part of the peer-relative decision.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: break queue read ties by mirror age

There's an unfair bias towards the first device in the pool of
candidate devices for queue-based read balancing: If two candidates are
mostly idle or share the same amount of inflight requests, the first
candidate always wins, no matter the history.

The age counter already provides history: Use the age counter to select
the device which has recently been ignored among the candidates if
there's a tie break on inflight IO requests. This distributes read
requests more evenly across all devices.

As part of this, a device with bad latency can be selected more often
now if it is last in the pool, or less often, if it is first in the
pool. This will be addressed by a later commit by adapting latency
history into the decision.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: move age io stat out of CONFIG_BTRFS_PER_DEVICE_IO_STATS

In a follow up commit, we will be using the age counter as a tie break
if two candidate devices currently have the same amount of inflight
requests pending. To use this effectively, the age counter must always
exist, no matter the io stats configuration.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add read latency health stats to per-device io read stats

Track a windowed average read latency per device and expose it through
the per-device read_stats file.

The health value is updated from block-layer read ios and wait-time
deltas in find_live_mirror(), independently of any adaptive read policy.
This lets read_stats diagnose latency-sick devices before
queue-adaptive selection is introduced.

The patch deliberately does not derive a slow flag from the per-device
stats. Whether a device is slow depends on the actual mirror candidates
for a read and must be decided later in the queue-adaptive selection
path.

v2:
- Leave missing mirrors unclassified instead of dereferencing a NULL
  block device on degraded arrays.
- Pace each device independently so one candidate cannot reset another
  candidate's sampling interval.
- Claim each update interval with cmpxchg() so concurrent readers cannot
  race or move the statistics checkpoint backwards.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add new Kconfig option for btrfs allocator hints

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add allocator_hint to disable allocation completely

This is useful where you want to prevent new allocations of chunks to
a set of multiple disks which are going to be removed from the pool.
This acts as a multiple `btrfs dev remove` on steroids that can remove
multiple disks in parallel without moving data to disks which would be
removed in the next round. In such cases, it will avoid moving the
same data multiple times, and thus avoid placing it on potentially bad
disks.

Thanks to @Zygo for the explanation and suggestion.

Link: https://github.com/kdave/btrfs-progs/issues/907#issuecomment-2520897104
Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add allocator_hint for no allocation preferred

This is useful where you want to avoid new allocations of chunks on a
disk which is going to be removed from the pool anyways, e.g. due to
bad blocks or because it's slow.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add allocator_hint mode

When this mode is enabled, the chunk allocation policy is modified as
follows:

Each disk may have a different tag:
- BTRFS_DEV_ALLOCATION_PREFERRED_METADATA
- BTRFS_DEV_ALLOCATION_METADATA_ONLY
- BTRFS_DEV_ALLOCATION_DATA_ONLY
- BTRFS_DEV_ALLOCATION_PREFERRED_DATA (default)

Where:
- ALLOCATION_PREFERRED_X means that it is preferred to use this disk
  for the X chunk type (the other type may be allowed when the space is
  low)
- ALLOCATION_X_ONLY means that it is used *only* for the X chunk type.
  This means also that it is a preferred choice.

Each time the allocator allocates a chunk of type X, first it takes the
disks tagged as ALLOCATION_X_ONLY or ALLOCATION_PREFERRED_X.

If the space is not enough, it uses also the disks tagged as
ALLOCATION_METADATA_ONLY.

If the space is not enough, it uses also the other disks, with the
exception of the one marked as ALLOCATION_PREFERRED_Y, where Y is the
other type of chunk (i.e. not X).

Co-authored-by: Goffredo Baroncelli <kreijack@inwind.it>
Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: export dev_item.type in /sys/fs/btrfs/<uuid>/devinfo/<devid>/type

Co-authored-by: Goffredo Baroncelli <kreijack@inwind.it>
Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: change the DEV_ITEM 'type' field via sysfs

v2: Adds a check to prevent modification while the file system is still mounting.

v3: Accept only valid allocation hint types

Todo:

- Transactions should not be triggered from sysfw writes, see:
  https://lore.kernel.org/linux-btrfs/20251213200920.1808679-1-kai@kaishome.de/

Link: https://github.com/kakra/linux/pull/36#issuecomment-3406301805
Reported-by: Eli Venter <eli@genedx.com>
Co-authored-by: Goffredo Baroncelli <kreijack@inwind.it>
Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add io read stats per device to devinfo

This adds read stats per device to devinfo to evaluate the effects of
different read policies better.

This adds a new file /sys/fs/btrfs/BTRFS-UUID/devinfo/ID/read_stats.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: add flags to give a hint to the chunk allocator

Add the following flags to give a hint about which chunk should be
allocated on which a disk.

The following flags are created:

- BTRFS_DEV_ALLOCATION_PREFERRED_DATA
  preferred data chunk, but metadata chunk allowed
- BTRFS_DEV_ALLOCATION_PREFERRED_METADATA
  preferred metadata chunk, but data chunk allowed
- BTRFS_DEV_ALLOCATION_METADATA_ONLY
  only metadata chunk allowed
- BTRFS_DEV_ALLOCATION_DATA_ONLY
  only data chunk allowed

Co-authored-by: Goffredo Baroncelli <kreijack@inwind.it>
Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
btrfs: scale adaptive queue guard with latency

The fixed queue guard allows a slow mirror to absorb overflow once
its in-flight count leads a healthier peer by more than two requests,
regardless of their latency difference. This can favor throughput over
tail latency when a mirror is substantially slower.

Scale the guard with the rounded-up ratio between the slow queue
winner and the selected healthy alternative. Clamp it between 2 and 16
requests to preserve the existing minimum while bounding how far load
can accumulate on the healthier mirror. Keep the minimum guard when the
alternative has no health sample yet.

The additional 64-bit division only runs when the normal queue winner
has already been classified as slow and a non-slow alternative exists.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
fixup! btrfs: add io read stats per device to devinfo

btrfs_devinfo_read_stats_show() calls part_stat_read() on device->bdev
unconditionally. A missing mirror on a degraded array has no bdev, so
the first read of that device's read_stats file dereferences a NULL
block device and crashes.

Guard both part_stat_read() calls behind device->bdev and fall back to
0 for read_ios/read_wait when it is absent. The existing zero check
before computing avg_wait already turns that into a report with
zeroed queue stats instead of a crash, and the remaining fields (age,
ignored, health avg/ago, vetoed, overflow) never depended on bdev to
begin with, so they keep reporting whatever history the device still
has.

v2:
- Guard part_stat_read() behind device->bdev to fix a NULL-pointer
  dereference when reading a missing mirror's stats on a degraded
  array.

Signed-off-by: Kai Krakow <kai@kaishome.de>

kakra committed:
fixup! btrfs: add read latency health stats to per-device io read stats

health_avg_ns, health_check_ios and health_check_wait are plain u64
fields updated via READ_ONCE()/WRITE_ONCE(). On 32-bit architectures a
64-bit load or store is not atomic, so a concurrent reader can observe
a torn value made up of half the old sample and half the new one. This
is advisory data only (read policy hints and sysfs diagnostics), but it
is inconsistent with the other per-device counters in the same struct
(last_io_age, stripe_ignored, health_vetoed, health_overflow), which
are already atomic64_t.

Make all three fields atomic64_t and switch their accessors to
atomic64_read()/atomic64_set(), matching the existing counters. On
64-bit architectures this compiles to the same plain load/store as
before. health_avg_jiffies and health_check_jiffies stay as they are:
they are unsigned long, already word-sized on every architecture, so
they cannot tear.

v3:
- Switch health_avg_ns, health_check_ios and health_check_wait to
  atomic64_t to avoid torn 64-bit reads/writes on 32-bit architectures,
  matching the other per-device counters already using atomic64_t.

Signed-off-by: Kai Krakow <kai@kaishome.de>

KeinNiemand commented:
I think I may have hit an edge case with allocator hints and RAID1C3 metadata placement.

My setup is:

* 4 HDDs + 1 NVMe in the same Btrfs filesystem
* Metadata profile: RAID1C3
* NVMe device ID 5 is set to type `1` (`PREFERRED_METADATA`)
* HDDs remain type `0`
* Goal is for one RAID1C3 metadata copy to live on the NVMe, with the other two copies remaining on HDDs

I confirmed the hint is set through:

```bash
cat /sys/fs/btrfs/<UUID>/devinfo/5/type
```

and the patched kernel/config are in use.

I then ran a metadata balance, but after processing about 35 metadata block groups the NVMe was still completely unused:

```text
/dev/nvme0n1p4, ID: 5
   Device size:           660.51GiB
   Unallocated:           660.51GiB
```

while metadata remained entirely on the HDDs:

```text
/dev/sdc1: Metadata,RAID1C3: 165.00GiB
/dev/sdb1: Metadata,RAID1C3:  82.00GiB
/dev/sda1: Metadata,RAID1C3:  83.00GiB
/dev/sdd1: Metadata,RAID1C3: 165.00GiB
NVMe:                         0.00GiB
```

The resulting distribution is especially interesting because the two HDDs with the most available space participate in essentially every metadata chunk, while the third copy is split almost exactly between the remaining two HDDs.

Looking through `gather_device_info()`, I noticed that devices are first sorted by `alloc_hint`, which correctly puts the type=1 NVMe first. But because RAID1C3 needs 3 devices and there is only one preferred-metadata device, the fallback hint group is also added.

After that, the code does:

```c
for (int i = 0 ; i < ndevs ; i++)
    devices_info[i].alloc_hint = 0;

sort(devices_info, ndevs, sizeof(struct btrfs_device_info),
     btrfs_cmp_device_info, NULL);
```

At that point the preference seems to be lost and the devices are sorted again by `max_avail`. In my setup the NVMe has less free space than several HDDs, so it appears to end up behind them and never becomes one of the actual RAID1C3 members.

Is this expected behavior, or is the intention that with one `PREFERRED_METADATA` device and RAID1C3 the allocator should produce something like:

```text
NVMe + HDD + HDD
```

for new metadata chunks?

If that is the intended behavior, I think this may be an edge case where the preferred device survives candidate-group selection but loses its preference before the final RAID device selection.

Zygo commented:
The preferences define the candidate groups, then the allocator selects within each group according to the original upstream algorithm, which for raid1* profiles, fills the drives with the most free space first.

So the behavior is expected--it causes problems with raid56 profiles and mixed data/metadata too.  The problem is that there's no way to express a preference for metadata without simultaneously expressing a dispreference for data.  In this case, you need to provide two metadata preference tiers before considering data, and that can't be done with the 4 metadata/data preferred/only selectors.

For this use case, you can assign 2 of the HDDs to PREFERRED_DATA and the other two to DATA_ONLY.  That will force the allocator to choose from exactly 3 drives for metadata, and one of them will be NVMe.  If the NVMe fills up, you'll need to switch the two DATA_ONLY drives back to PREFERRED_DATA, so this setup will require ongoing monitoring.

Alternatively, implement 4-bit-wide independent preference priority for metadata and data, instead of the 1-bit-wide scheme currently implemented.  That would allow defining one metadata tier with NVMe+HDD1+HDD2, then a second tier with NVMe+HDD1+HDD2+HDD3+HDD4 to handle the case where NVMe fills up.  The data tiers would then be just one tier, HDD1+HDD2+HDD3+HDD4.  It wouldn't be compatible with the current scheme, unless the preferences were stored somewhere else and the patch implemented a legacy configuration translator to the 4-bit scheme.

KeinNiemand commented:
I’m a bit confused why this requires either the `DATA_ONLY` workaround or a new 4-bit priority scheme.

My actual goal is simpler: for RAID1C3 metadata, I want one preferred-metadata device to take one copy, then let the normal allocator choose the other two copies from the HDDs.

So with:

```text
NVMe = PREFERRED_METADATA
HDD1-4 = PREFERRED_DATA
```

I would expect something like:

1. Take the preferred-metadata NVMe.
2. We still need 2 devices to satisfy `devs_min`.
3. From the lower-priority HDD group, pick only the 2 devices with the highest `max_avail`.
4. Hand those 3 devices to the normal downstream allocator, still sorted by `max_avail`.

That would give:

```text
NVMe + 2 HDDs
```

while still allowing the two HDD copies to move around normally depending on free space.

Is there a reason the allocator has to add the entire lower-priority group instead of only enough devices from that group to satisfy `devs_min`?

That part is what I’m missing.

Zygo commented:
> For each metadata chunk, I want one copy to prefer the NVMe, and the other two copies to go to whichever two HDDs the normal allocator would choose based on free space. I do **not** want to prefer two specific HDDs.

This cannot be expressed in the current schema.  When you say "PREFERRED_DATA", you are implicitly saying "reject metadata", so all the preferred_data drives are excluded from the first allocation pass.  The second pass includes _all_ of the preferred_data drives, because there's currently no way to express which of those you want the allocation to be satisfied from.

What you want here is "select _some_ from pool A, then select _more_ from a pool B", but the downstream interface is "select _all_ from _the_ pool."  It takes only one set of devices and does all-or-nothing allocation over that set.

The 4-bit preferences allow userspace to do the selection in the device sort, so "the" pool gets larger in multiple minimal steps.  That solves the "put metadata on the NVMe" part of your request, but not the "pick the best 2 HDDs out of 4 HDDs" part.  That piece can be done by userspace, e.g. once a day, a script checks the amount of unallocated space on the drives, and adjusts the preference to balance out the space usage.  This is only needed if you really want all 4 HDDs to have metadata; otherwise, the data allocator will easily compensate if only two spinning drives have metadata.  If your data drives are different sizes, there might be only one preference ordering anyway.

> So I'm confused about which of these is the issue:
> 
> 1. Did you understand my goal as wanting a fixed preferred set like `NVMe + HDD1 + HDD2`?

That's the _closest_ result to what you wanted that you can get from the current patches.

> 2. Or is there some allocator reason why this cannot work by first taking devices from the higher-priority group, then, if that does not satisfy `devs_min`, taking only as many devices as still needed from the next group using the normal `max_avail` ordering?

> What I'm missing is why the preference groups need to be effectively all-or-nothing here, instead of filling the required mirror members tier by tier.

This would only do the right thing in this use case--and even then, only when the sort function can figure out which drives have free space.  It would be the wrong thing for other cases.

For striped profiles, the _widest_ candidate set is the best answer:  raid0/5/6 currently fails to pick the second-tier devices when they would be wanted.  In this case, we'd want to skip the first tier entirely (that tier would exist only to prevent data from being allocated on NVMe), and go straight to allocating on all 5 drives at once.  That's the opposite of what's needed for your use case.

So to get what you want, the preferences could no longer be a simple device sort before handing off to the profile-specific allocator.  Every profile-specific allocator would have to understand the preferences and do its own device sort, because they would need to select _which_ of the drives from the larger pool to add to the smaller one.  They would _also_ need to know when to override that preference, e.g. when the smaller NVMe is full and can no longer be part of the candidate group.  That would double the size of this patch, or maybe it'd be easier to start over by writing a new patch, since it's a refactor and we wouldn't need the _existing_ sort code any more.

This is something of an outlier case for allocation preferences.  Normally, if a user has preferred metadata drives, they have enough of each type to allocate the metadata profile entirely on the preferred type, and only resort to filling in HDDs when they've completely run out of all NVMe space.  This covers setups with e.g. 2 metadata-only SSDs and 6 data-preferred HDDs to allow metadata overflow (with no block group using both NVMe and HDD at the same time), or the 1 metadata-preferred NVME + 1 data-preferred HDD configuration found in laptops, where there's no "extra" devices to be allocated out of profile order.  It doesn't extend as far as overriding the allocation profiles' normal allocation rules.

kakra commented:
Thanks, both, @Zygo and @KeinNiemand. I traced the current 6.18 chunk allocation path more carefully. Zygo, I agree with your point for wide striped profiles: RAID0/10/5/6 cannot generally use the same tier-filling rule as fixed-width mirrored profiles. Preserving strict tier order there could reduce the stripe width or include/exclude fallback devices at the wrong time.

There may, however, be a narrower solution specifically for RAID1, RAID1C3 and RAID1C4 without teaching every profile allocator about hints.

For these profiles, the existing allocator already has a fixed device count:

```
RAID1   devs_min/devs_max = 2/2
RAID1C3 devs_min/devs_max = 3/3
RAID1C4 devs_min/devs_max = 4/4
```

After `gather_device_info()`, `decide_stripe_size()` limits `ctl->ndevs` to `devs_max`, and `create_chunk()` consumes exactly the `first ctl->ndevs` entries without reordering them.

The current problem appears to be the second sort in `gather_device_info()`: once enough complete hint groups have been collected to satisfy `devs_min`, the hints are cleared and the whole candidate set is sorted globally by free space. The later `devs_max` truncation therefore no longer sees the preference order.

A possible narrowly scoped implementation would be:

1. Keep the current hint-group collection unchanged.
2. For RAID0/10/5/6, retain the current behavior exactly.
3. Only for `BTRFS_BLOCK_GROUP_RAID1_MASK`, preserve the hint-first ordering until `ctl->ndevs` has been limited to the fixed mirror count.
4. Continue ordering devices by the normal `max_avail`/`total_avail` criteria within each hint tier.
5. After selecting the first 2/3/4 devices, sort only that selected subset by free space before calculating the stripe size.

The last step is important because `decide_stripe_size_regular()` currently assumes that `devices_info[ctl->ndevs - 1]` has the smallest `max_avail`. Keeping the hint order all the way into that function would violate this invariant when a preferred NVMe has less free space than the selected fallback HDDs.

For example, RAID1C3 with one preferred NVMe and four fallback HDDs would first produce:

- NVMe, best HDD, second-best HDD, third HDD, fourth HDD

The existing fixed `devs_max = 3` selection would retain:

- NVMe, best HDD, second-best HDD

Only those three would then be reordered for stripe-size calculation. A preferred device with no usable extent would continue to be skipped during gathering, so fallback should still work as it does now.

This would deliberately not change the semantics of RAID0/10/5/6. It also would not introduce a new on-disk format or promise that every allocation-preference use case can be expressed.

I have not implemented or tested this yet, so this is only a proposed direction. @Zygo, do you see another downstream invariant or profile interaction that would make this RAID1-only approach incorrect? In particular, is there a reason the existing fixed `devs_max` selection cannot serve as the boundary between preference-based member selection and the normal free-space ordering?

Zygo commented:
> There may, however, be a narrower solution specifically for RAID1, RAID1C3 and RAID1C4 without teaching every profile allocator about hints.

This is "add a bridge layer that knows how all the profiles work, then use it to massage the data until existing downstream functions can consume it," which isn't very different in terms of complexity from "build knowledge of preferences into every downstream profile."  It still has custom behavior for every profile.

It might even be a little worse, because now there's _two_ pieces of code to maintain in lockstep, where a refactoring (move the preference sort downstream) would be only _one_.

In the LLM age...generate both versions, see which one is less ugly, or which one gets to passing tests first.

You haven't mentioned what happens to single and dup profiles.  I think they go into the "unchanged" bucket--since their min_devs == max_devs == 1, their allocations can only be all-or-nothing, so there's no need to handle a partially filled block group.  If an -only device is full, the fallback already moves on to the next tier.

This doesn't solve the raid56 stripe width problem.  That requires a raid0/5/6/10-specific change going the other way, making data stripes as wide as possible.  It effectively merges data_only and data_preferred into the same tier so no allocation is attempted on data_only by itself.  If that fails, then add metadata_preferred to the pool as a last resort.

I'm not sure which bucket raid10 falls into.  It wants widest stripe like raid0, but uses size-sorting to break ties and equalize space usage on odd-numbered array sizes like raid1.

Unless I missed it, I don't see where you handle the case of "most preferred device is full and allocation there is not possible."  Or are completely-full devices already excluded before preference sorting?  Actually, the answer to that question doesn't matter, because if it's 'no', change the code so it is 'yes.'  Exclude completely-full devices before preference sorting, then when the NVMe fills up, it stops being a preferred device, the top-3-devices sort finds the top 3 HDD's, and everything downstream adapts naturally.

kakra commented:
@KeinNiemand I'm reviewing your patch with my multi-AI-agent pipeline, and currently it looks good and safe to use. I made just a minor adjustment to include @Zygo's suggestions.

kakra commented:
**Update:** I force-pushed the branch after folding the remaining fixup commits into their respective patches and preserving the relevant changes as version notes in the commit messages. Existing commit hashes have therefore changed.

The allocator-hints implementation now also handles fixed-width mirrored profiles as discussed:

- RAID1, RAID1C3 and RAID1C4 preserve allocator-hint priority when their required mirrors span multiple hint tiers.
- Preferred devices are selected first, ordered by the normal allocator criteria within their tier.
- Only the required number of candidates is retained from the final tier.
- The selected devices are then sorted normally for the existing stripe-size calculation.
- RAID0, RAID10, RAID5 and RAID6 retain their previous candidate-group behavior.

For example, RAID1C3 metadata allocation can now select:

> preferred NVMe + preferred NVMe + best fallback HDD

instead of losing the smaller preferred devices when all admitted candidates are sorted globally by free space.

The existing type values and sysfs interface remain unchanged.

The documentation now also makes the current scope and limitations explicit:

- Mixed data/metadata block groups intentionally ignore allocator hints. They cannot separate data and metadata placement and are outside the intended tiering use case.
- `btrfs device replace` does not transfer the allocation hint to the target device. The hint must currently be reapplied after replacement.
- Sysfs hint updates are not serialized with concurrent chunk allocation, so an allocation racing with an update may observe either the old or new value.
- A nearly full preferred device may constrain the resulting chunk size. This already happened when the preferred tier contained enough devices; the new selection behavior extends it to fixed-width mirrors spanning multiple tiers.

A later allocator-hints design is expected to replace the overloaded device `type` field with dedicated transactional structures. That should also provide a proper solution for device replacement and update synchronization.

The fixed-width mirror change received an additional review focused on candidate selection and stripe sizing. No unintended behavior change was found for the other RAID profiles. The relevant Btrfs objects also build cleanly with `W=1`.

Thanks to @KeinNiemand for reporting the RAID1C3 case and providing the initial implementation, and to @Zygo for the detailed allocator-semantics feedback.
